heat/core/linalg/bcg.py

```diff
@@ -100,59 +107,338 @@ def bi_diagonalize(A, overwrite_arr=True):
+    There is a slight difference in the way band is partitioned into block columns Dj,Ej for the cases where m>=n & n>m.
+
+
+
+    To change the serial algorithm into an Efficient parallel program, We should distribute the data and the computational work to multiple processors.
+
+    The band form is further reduced to the final condensed form using the bulge chasing technique.
+    This procedure annihilates the extra off-diagonal elements by chasing the created
+    fill-in elements down to the bottom right side of the matrix using successive
+    orthogonal transformations at each sweep.
+
+
+
+    Note: We will do the bulge chasing using Halos.
+
+    Second Stage of reduction: Band matrix to a real bidiagonal matrix. This uses a bulge chasing algorithm.
+
+
     """
     if overwrite_arr:
         arr = A
         # The input matrix is overwritten with the result, i.e it will be changed to a bidiagonal matrix.
     else:
         arr = A.copy()
-        # The input matrix is not overwritten, i.e a new matrix "arr" which is a copy of input matrix, will be changed to a bidiagonal matrix.
+        # The input matrix is not overwritten, i.e a new matrix "arr" which is a copy of the input matrix, will be changed to a bidiagonal matrix.
 
     m, n = arr.shape
+    k = min(m, n)
+    # k is the minimum of m and n
+
+    # Find the width of the diagonal of the input matrix.
+
+    b = 0
+    bl = 0
+    bu = 0
+    row_0 = arr[0, 0:]
+    col_0 = arr[0:, 0]
+
+    for i in range(1, len(col_0)):
+        if col_0[i] != 0:
+            bl += 1
+
+    for i in range(1, len(row_0)):
+        if row_0[i] != 0:
+            bu += 1
+
+    b = bl + bu
+
+    # print("b: ", b)
+
+    U1, vt1 = ht.eye(m, dtype=ht.float64), ht.eye(n, dtype=ht.float64)
+    # U1 is an identity matrix of size m x m, vt1 is an identity matrix of size n x n
+
+    # print(arr)
+
+    for i in range(k):
+
+        if m >= n:
+            if rank == 0 and i < k - 1:
+                # print(f"at {i} arr = ", arr)
+                D1 = arr[i : i + 1 + bl, i]
+                v_left, tau_left = gen_house_vec(D1)
+                Uj = apply_house_left(D1, v_left, tau_left, U1, D1.shape[0], i)
+                arr[i : i + 1 + bl, i] = D1
+
+                comm.send(arr, dest=1, tag=1)
+                comm.send(Uj, dest=1, tag=2)
+
+                req2 = comm.irecv(source=1, tag=3)
+                arr = req2.wait()
+
+            else:
+                # print(f"at {i} arr = ", arr)
+                D1 = arr[i : i + 1 + bl, i]
+                v_left, tau_left = gen_house_vec(D1)
+                Uj = apply_house_left(D1, v_left, tau_left, U1, D1.shape[0], i)
+                arr[i : i + 1 + bl, i] = D1
+
+            if i < k - 1:
+                for j in range(2, k):
+                    # print("j is = ", j)
+
+                    if j == 2:
+                        if rank == 1:
```

Comment:
these can be combined into a single statement

Reply:
Done sir, 👍